portability-core/src/main/java/org/dataportabilityproject/cloud/google/GooglePersistentKeyValueStore.java

```diff
@@ -46,12 +46,97 @@
 
   private final Datastore datastore;
 
+  // Current transaction for completing atomic operations. Only one atomic transaction may be
+  // executed at a time. Null if we are not currently completing a transaction.
+  private Transaction currentTransaction = null;
```

Comment:
I think there is a potential for java concurrency issues.  Multiple threads could make calls to an instance of this class and receive different state.

One solution we could try is to to expose a simple interface for atomic updates that will implement the underlying datastores transactions:
https://cloud.google.com/datastore/docs/concepts/transactions#datastore-transactional-update-java

There may be others.

Reply:
Yeah, was wondering about concurrency issues. In the API, right?

What do you think about instead of the current transaction, I use a concurrent map of id -> transaction, and let the users of this class pass the transaction ID?